### PR TITLE
move initial-work to master

### PR DESCRIPTION
**Pull request by gemmellr**

### CODE REVIEW

File: src/main/asciidoc/java/index.adoc [hunk 1]
```diff
@@ -0,0 +1,89 @@
+= Vert.x Proton
+
+This component faciliates AMQP integrations for Vert.x by providing a thin wrapper around the
```
Comment (vietj) [inline src/main/asciidoc/java/index.adoc]: facilitates

File: src/main/java/io/vertx/proton/ProtonClient.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
+package io.vertx.proton;
```
Comment (vietj) [inline src/main/java/io/vertx/proton/ProtonClient.java]: headers

File: src/main/java/io/vertx/proton/ProtonClientOptions.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package io.vertx.proton;
+
+import io.vertx.core.net.NetClientOptions;
+
+public class ProtonClientOptions extends NetClientOptions {
```
Comment (vietj) [inline src/main/java/io/vertx/proton/ProtonClientOptions.java]: doc

File: src/main/java/io/vertx/proton/ProtonClientOptions.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  }
+
+  /**
+   * Set a restricted mechanism(s) that the client may use during the SASL negotiation. If null or empty argument is
+   * given, no restriction is applied and any supported mechanism can be used.
+   *
+   * @param mechanisms
+   *          the restricted mechanism(s) or null to clear the restriction.
+   */
+  public void setAllowedSaslMechanisms(final String... saslMechanisms) {
```
Comment (vietj) [inline src/main/java/io/vertx/proton/ProtonClientOptions.java]: fluent return

File: src/main/java/io/vertx/proton/ProtonLink.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Handler;
+import org.apache.qpid.proton.amqp.transport.ErrorCondition;
+import org.apache.qpid.proton.amqp.transport.Source;
+import org.apache.qpid.proton.amqp.transport.Target;
+
+/**
+ * @author <a href="http://tfox.org">Tim Fox</a>
+ */
+public interface ProtonLink<T extends ProtonLink> {
```
Comment (vietj) [inline src/main/java/io/vertx/proton/ProtonLink.java]: should be `<T extends ProtonLink<T>>`

File: src/main/java/io/vertx/proton/ProtonSender.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+   */
+  ProtonDelivery send(byte[] tag, Message message, Handler<ProtonDelivery> onUpdated);
+
+  /**
+   * Gets whether the senders outgoing send queue is full, i.e. there is currently no credit to send and send
+   * operations will actually buffer locally until there is.
+   *
+   * @return whether the send queue is full
+   */
+  boolean sendQueueFull();
```
Comment (vietj) [inline src/main/java/io/vertx/proton/ProtonSender.java]: isSendQueuFull ?

File: src/main/java/io/vertx/proton/ProtonServerOptions.java [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package io.vertx.proton;
+
+import io.vertx.core.net.NetServerOptions;
+
+public class ProtonServerOptions extends NetServerOptions {
```
Comment (vietj) [inline src/main/java/io/vertx/proton/ProtonServerOptions.java]: Override return type indeed and return covariant type

File: src/main/java/io/vertx/proton/ProtonSession.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+   */
+  ProtonSession close();
+
+  /**
+   * Sets the incoming capacity in bytes, used to govern session-level flow control.
+   *
+   * @param capacity
+   *          capacity in bytes
+   */
+  void setIncomingCapacity(int capacity);
```
Comment (vietj) [inline src/main/java/io/vertx/proton/ProtonSession.java]: fluent

File: src/main/java/io/vertx/proton/ProtonSession.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+  int getIncomingCapacity();
+
+  /**
+   * Sets the local ErrorCondition object.
+   *
+   * @param condition
+   *          the condition to set
+   * @return the link
+   */
+  void setCondition(ErrorCondition condition);
```
Comment (vietj) [inline src/main/java/io/vertx/proton/ProtonSession.java]: fluent

File: src/main/java/io/vertx/proton/ProtonSender.java (hunk 6 above)
Comment (vietj) [inline src/main/java/io/vertx/proton/ProtonSender.java]: actually this is WriteStream so don't bother

File: src/test/java/io/vertx/proton/FutureHandler.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+  private T result() throws ExecutionException {
+    synchronized (this) {
+      if (exception != null) {
+        throw exception;
+      }
+      return result;
+    }
+  }
+
+  public T get(long timeout, TimeUnit unit) throws InterruptedException, TimeoutException, ExecutionException {
```
Comment (vietj) [inline src/test/java/io/vertx/proton/FutureHandler.java]: rename to getBlocking

File: src/main/java/io/vertx/proton/sasl/impl/ProtonSaslExternalImpl.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
+package io.vertx.proton.sasl.impl;
+
+public class ProtonSaslExternalImpl extends ProtonSaslMechanismImpl {
```
Comment (vietj) [inline src/main/java/io/vertx/proton/sasl/impl/ProtonSaslExternalImpl.java]: how making it a singleton ?

File: src/main/java/io/vertx/proton/sasl/ProtonSaslMechanism.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+  byte[] getChallengeResponse(byte[] challenge) throws SaslException;
+
+  /**
+   * Sets the user name value for this Mechanism. The Mechanism can ignore this value if it does not utilize user name
+   * in it's authentication processing.
+   *
+   * @param username
+   *          The user name given.
+   */
+  void setUsername(String username);
```
Comment (vietj) [inline src/main/java/io/vertx/proton/sasl/ProtonSaslMechanism.java]: fluent

File: src/main/java/io/vertx/proton/sasl/ProtonSaslMechanism.java [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+  String getUsername();
+
+  /**
+   * Sets the password value for this Mechanism. The Mechanism can ignore this value if it does not utilize a password
+   * in it's authentication processing.
+   *
+   * @param username
+   *          The user name given.
+   */
+  void setPassword(String username);
```
Comment (vietj) [inline src/main/java/io/vertx/proton/sasl/ProtonSaslMechanism.java]: fluent

File: src/test/java/io/vertx/proton/FutureHandler.java (hunk 10 above)
Comment (gemmellr) [inline src/test/java/io/vertx/proton/FutureHandler.java]: Those methods are implementing the java.util.concurrent.Future interface and so can't really be renamed. It is test-only code that the tests use specifically to block, is it an issue to leave it as is?

File: src/main/java/io/vertx/proton/sasl/impl/ProtonSaslExternalImpl.java (hunk 11 above)
Comment (gemmellr) [inline src/main/java/io/vertx/proton/sasl/impl/ProtonSaslExternalImpl.java]: The superclass has state so making it a singleton seems a bit off, I think I'd prefer it the way it is.

**Comment by gemmellr** [comment]
Thanks for the review @vietj. I have made updates to address most of the comments, and commented on a couple. Can you (or anyone else :) ) give things another look?

Thanks,
Robbie

File: src/test/java/io/vertx/proton/FutureHandler.java (hunk 10 above)
Comment (vietj) [inline src/test/java/io/vertx/proton/FutureHandler.java]: ok, I didn't spotted that.

File: src/main/java/io/vertx/proton/sasl/impl/ProtonSaslExternalImpl.java (hunk 11 above)
Comment (vietj) [inline src/main/java/io/vertx/proton/sasl/impl/ProtonSaslExternalImpl.java]: ok, I thought iwas overriding all the state returning methods, my bad.

**Comment by vietj** [comment]
looks good @gemmellr you can merge

**Comment by gemmellr** [comment]
Thanks @vietj, done.

**Comment by vietj** [comment]
awesome!

**Pull request merged**